File: backend/src/main/java/io/ntf/api/infrastructure/SecurityConfiguration.java
package io.ntf.api.infrastructure;

import io.ntf.api.infrastructure.jwt.JwtWebSecurityConfigurer;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.web.reactive.EnableWebFluxSecurity;
import org.springframework.security.config.web.server.ServerHttpSecurity;
import org.springframework.security.web.server.SecurityWebFilterChain;

import static org.springframework.http.HttpMethod.*;

@EnableWebFluxSecurity
@Configuration
public class SecurityConfiguration {

  private final Auth0ConfigurationProperties auth0ConfigurationProperties;

  @Autowired
  public SecurityConfiguration(Auth0ConfigurationProperties auth0ConfigurationProperties) {
    this.auth0ConfigurationProperties = auth0ConfigurationProperties;
  }

  @Bean
  public SecurityWebFilterChain securityWebFilterChain(ServerHttpSecurity http) {
    String issuer = auth0ConfigurationProperties.getIssuer();
    String audience = auth0ConfigurationProperties.getAudience();

    //@formatter:off
    JwtWebSecurityConfigurer.forRS256(audience, issuer)
      .configure(http)
      .csrf().disable()
      .authorizeExchange()
        .pathMatchers(POST, "/activity").hasAuthority("start:activity")
        .pathMatchers(POST, "/activity/stop").hasAuthority("stop:activity")
        .pathMatchers(GET, "/activity/running").hasAuthority("read:activities")
        .pathMatchers(PUT, "/activity/{id}").hasAuthority("update:activity")
        .pathMatchers(DELETE, "/activity/{id}").hasAuthority("delete:activity")
        .pathMatchers(DELETE, "/activity/{id}").hasAuthority("delete:activity")
        .pathMatchers(GET, "/activities").hasAuthority("read:activities")
        .pathMatchers(GET, "/logs").hasAuthority("read:logs")
        .pathMatchers(POST, "/log").hasAuthority("create:log")

        .pathMatchers(GET, "/statistics/overtime").hasAuthority("read:overtime")

        .pathMatchers(OPTIONS).permitAll();
    //@formatter:off

    return http.build();
  }

}

File: backend/src/main/kotlin/io/ntf/api/infrastructure/SecurityConfiguration.kt
package io.ntf.api.infrastructure

import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.HttpMethod.*
import org.springframework.security.config.annotation.web.reactive.EnableWebFluxSecurity
import org.springframework.security.config.web.server.ServerHttpSecurity
import org.springframework.security.oauth2.core.DelegatingOAuth2TokenValidator
import org.springframework.security.oauth2.jwt.*
import org.springframework.security.web.server.SecurityWebFilterChain

@EnableWebFluxSecurity
@Configuration
@EnableConfigurationProperties(Auth0ConfigurationProperties::class)
class SecurityConfiguration(private val auth0ConfigurationProperties: Auth0ConfigurationProperties) {

  @Bean
  fun securityWebFilterChain(http: ServerHttpSecurity): SecurityWebFilterChain =
    http
      .authorizeExchange()
      .pathMatchers(OPTIONS).permitAll()
//TODO fix scopes
      .pathMatchers(POST, "/activity").hasAuthority("SCOPE_start:activity")
      .pathMatchers(POST, "/activity/stop").hasAuthority("SCOPE_stop:activity")
      .pathMatchers(GET, "/activity/running").hasAuthority("SCOPE_read:activities")
      .pathMatchers(PUT, "/activity/{id}").hasAuthority("SCOPE_update:activity")
      .pathMatchers(DELETE, "/activity/{id}").hasAuthority("SCOPE_delete:activity")
      .pathMatchers(GET, "/activities").hasAuthority("SCOPE_read:activities")
      .pathMatchers(GET, "/activities/{id}").hasAuthority("SCOPE_read:activities")
      .pathMatchers(GET, "/activities/{from}/{to}").hasAuthority("SCOPE_read:activities")
      .pathMatchers(GET, "/activities/tags").hasAuthority("SCOPE_read:activities")
      .pathMatchers(POST, "/activity/repeat").hasAuthority("SCOPE_start:activity")

      .pathMatchers(GET, "/metrics").hasAuthority("SCOPE_read:metrics")
      .pathMatchers(POST, "/metrics").hasAuthority("SCOPE_create:metrics")
      .pathMatchers(GET, "/metrics/{id}").hasAuthority("SCOPE_read:metrics")
      .pathMatchers(DELETE, "/metrics/{id}").hasAuthority("SCOPE_delete:metrics")
      .pathMatchers(POST, "/metrics/{id}").hasAuthority("SCOPE_update:metric")
      .pathMatchers(GET, "/metrics/{id}/config").hasAuthority("SCOPE_read:metrics")
      .and()
      .oauth2ResourceServer().jwt().and().and().build()

  @Bean
  fun jwtDecoder(): ReactiveJwtDecoder {
    val issuer = auth0ConfigurationProperties.issuer
    val audience = auth0ConfigurationProperties.audience

    val jwtDecoder = ReactiveJwtDecoders.fromOidcIssuerLocation(issuer) as NimbusReactiveJwtDecoder

    val audienceValidator = AudienceValidator(audience)
    val withIssuer = JwtValidators.createDefaultWithIssuer(issuer)
    val withAudience = DelegatingOAuth2TokenValidator(withIssuer, audienceValidator, JwtTimestampValidator())

    jwtDecoder.setJwtValidator(withAudience)

    return jwtDecoder
  }

}
